lib/vast/Dialect/HighLevel/HighLevelOps.cpp

```diff
@@ -34,6 +37,134 @@ namespace vast::hl
 
     using FoldResult = mlir::OpFoldResult;
 
+    //===----------------------------------------------------------------------===//
+    // FuncOp
+    //===----------------------------------------------------------------------===//
+
+    // This function is adapted from CIR:
+    //
+    // Verifies linkage types, similar to LLVM:
+    // - functions don't have 'common' linkage
+    // - external functions have 'external' or 'extern_weak' linkage
+    LogicalResult FuncOp::verify() {
+        auto linkage = getLinkage();
+        constexpr auto common = GlobalLinkageKind::CommonLinkage;
+        if (linkage == common) {
+            return emitOpError() << "functions cannot have '"
+                << stringifyGlobalLinkageKind(common)
+                << "' linkage";
+        }
+
+        if (isExternal()) {
+            constexpr auto external = GlobalLinkageKind::ExternalLinkage;
+            constexpr auto weak_external = GlobalLinkageKind::ExternalWeakLinkage;
+            if (linkage != external && linkage != weak_external) {
+                return emitOpError() << "external functions must have '"
+                    << stringifyGlobalLinkageKind(external)
+                    << "' or '"
+                    << stringifyGlobalLinkageKind(weak_external)
+                    << "' linkage";
+            }
+            return mlir::success();
+        }
+        return mlir::success();
+    }
+
+    void add_arg_attrs(Builder &bld, State &st, llvm::ArrayRef< mlir::DictionaryAttr > arg_attrs) {
+        auto non_empty_attrs = [] (mlir::DictionaryAttr attrs) {
+            return attrs && !attrs.empty();
+        };
+
+        // Convert the specified array of dictionary attrs (which may have null
+        // entries) to an ArrayAttr of dictionaries.
+        auto get_array_attr = [&] (llvm::ArrayRef< mlir::DictionaryAttr > dict_attrs) {
+            llvm::SmallVector< Attribute > attrs;
+            for (auto &dict : dict_attrs) {
+                attrs.push_back(dict ? dict : bld.getDictionaryAttr({}));
+            }
+            return bld.getArrayAttr(attrs);
+        };
+
+        // Add the attributes to the function arguments.
+        if (llvm::any_of(arg_attrs, non_empty_attrs)) {
+            st.addAttribute(mlir::function_interface_impl::getArgDictAttrName(), get_array_attr(arg_attrs));
+        }
+    }
+
+    ParseResult parseFunctionSignaruteAndBody(
+        Parser &parser, Attribute &funcion_type, mlir::NamedAttrList &attr_dict, Region &body
+    ) {
+        llvm::SmallVector< Parser::Argument, 8 > arguments;
```

Comment:
How are these magic constants derived? Maybe add a comment that says how they came to be.